p4-16/psa/examples/psa-example-incremental-checksum.p4

```diff
@@ -0,0 +1,202 @@
+                         out psa_parser_output_metadata_t ostd)
+{
+    state start {
+        buffer.extract(parsed_hdr.ethernet);
+        transition select(parsed_hdr.ethernet.etherType) {
+            0x0800: parse_ipv4;
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        buffer.extract(parsed_hdr.ipv4);
+        verify(parsed_hdr.ipv4.ihl == 5, error.UnhandledIPv4Options);
+        transition select(parsed_hdr.ipv4.protocol) {
+            6: parse_tcp;
+            default: accept;
+        }
+    }
+    state parse_tcp {
+        buffer.extract(parsed_hdr.tcp);
+        transition accept;
+    }
+}
+
+// BEGIN:Incremental_Checksum_Table    
+control ingress(inout headers hdr,
+                inout metadata user_meta,
+                PacketReplicationEngine pre,
+                in  psa_ingress_input_metadata_t  istd,
+                out psa_ingress_output_metadata_t ostd) {
+    action drop() {
+      ingress_drop(ostd);
+    }
+    action forward(PortId_t port, bit<32> srcAddr) {
+      user_meta.fwd_metadata.old_srcAddr = hdr.ipv4.srcAddr;
+      hdr.ipv4.srcAddr = srcAddr;
+      send_to_port(ostd, port);      
+    }
+    table route {
+        key = { hdr.ipv4.dstAddr : lpm; }
+        actions = {
+          forward;
+          drop;
+        }
+    }
+    apply {
+        if(hdr.ipv4.isValid()) {
+          route.apply();
+        }
+    }
+}
+// END:Incremental_Checksum_Table        
+
+parser EgressParserImpl(packet_in buffer,
+                        out headers parsed_hdr,
+                        inout metadata user_meta,
+                        in psa_egress_parser_input_metadata_t istd,
+                        out psa_parser_output_metadata_t ostd)
+{
+    state start {
+        transition accept;
+    }
+}
+
+control egress(inout headers hdr,
+               inout metadata user_meta,
+               BufferingQueueingEngine bqe,
+               in  psa_egress_input_metadata_t  istd,
+               out psa_egress_output_metadata_t ostd)
+{
+    apply { }
+}
+
+// BEGIN:Incremental_Checksum_Example
+control DeparserImpl(packet_out packet, inout headers hdr, in metadata user_meta) {
+    InternetChecksum() ck;
+    bit<16> hdrChecksum;
+    apply {
+        // Update IPv4 checksum
+        ck.clear();
+        ck.update({ hdr.ipv4.version,
+            hdr.ipv4.ihl,
+            hdr.ipv4.diffserv,
+            hdr.ipv4.totalLen,
+            hdr.ipv4.identification,
+            hdr.ipv4.flags,
+            hdr.ipv4.fragOffset,
+            hdr.ipv4.ttl,
+            hdr.ipv4.protocol,
+            //hdr.ipv4.hdrChecksum, // intentionally leave this out
+            hdr.ipv4.srcAddr,
+            hdr.ipv4.dstAddr });
+        hdrChecksum = ck.get();
+        // Update TCP checksum
+       ck.clear();
+       ck.remove(hdr.tcp.checksum);
+       ck.remove(user_meta.fwd_metadata.old_srcAddr);
+       ck.remove(hdr.ipv4.hdrChecksum);
```

Comment:
The TCP pseudo-header includes the IPv4 source and dest address, but not the IPv4 header checksum, so this remove call, and the ck.update(hdrChecksum) call below, should be removed.  RFC 793 is a good reference for the contents of the TCP pseudo-header.

Reply:
I am going to go ahead and approve and merge in these changes, and then make the small fix suggested above on top of that.